driver/src/price_finding/stablex_naive_solver.rs

```diff
@@ -0,0 +1,208 @@
+use crate::price_finding::error::PriceFindingError;
+use crate::price_finding::PriceFinding;
+use crate::util;
+use dfusion_core::models::{AccountState, Order, Solution};
+use std::cmp::Ordering;
+
+/// A naive solver for the stableX contract. It works almost identically to the
+/// Snapp naive solver except it has a few extra requirements and it doesn't try
+/// to be as optimal.
+///
+/// The solver works by finding the first pair of orders which matching sell and
+/// buy tokens, where one of the traded tokens is the fee token, whose users have
+/// enough balance to fully fill, and that has a viable solution.
+///
+/// The solution is computed by taking the price of the larger order and and
+/// completely filling the smaller order. A large order is defined by having
+/// more non-fee token volume. Account balances are checked for each solution.
+pub struct StableXNaiveSolver;
+
+impl PriceFinding for StableXNaiveSolver {
+    fn find_prices(
+        &self,
+        orders: &[Order],
+        accounts: &AccountState,
+    ) -> Result<Solution, PriceFindingError> {
+        Ok(ipairs(orders)
+            .filter(|&(i, j)| {
+                orders[i].buy_token == orders[j].sell_token
+                    && orders[i].sell_token == orders[j].buy_token
+            })
+            .filter(|&(i, _)| orders[i].buy_token == 0 || orders[i].sell_token == 0)
+            .filter(|&(i, j)| {
+                has_enough_balance(&orders[i], accounts) && has_enough_balance(&orders[j], accounts)
+            })
+            .filter_map(|(i, j)| compute_solution(orders, i, j))
+            .nth(0)
+            .unwrap_or_else(|| Solution::trivial(orders.len())))
+    }
+}
+
+const ETH: u128 = 1_000_000_000_000_000_000u128;
+const FEE_DENOM: u128 = 1000;
+
+/// Returns an iterator over all unordered pairs of indices in slice.
+fn ipairs<T>(slice: &[T]) -> impl Iterator<Item = (usize, usize)> {
+    let len = slice.len();
+    (0..len - 1)
+        .map(move |i| (i + 1..len).map(move |j| (i, j)))
+        .flatten()
+}
+
+/// Returns true if the owner of an order has enough balance to fully fill the
+/// order; false otherwise.
+fn has_enough_balance(order: &Order, accounts: &AccountState) -> bool {
```

Comment:
It would be nicer to just adapt the order to the available balance, instead of throwing it out. But, since this is extra effort, I am okay with letting it as it is

Reply:
I want to eventually only have one naive solver. This is just an interim solution so the e2e tests for stableX.